packages/styled-base/types/index.d.ts

```diff
@@ -33,50 +35,45 @@ export interface StyledComponent<InnerProps, StyleProps, Theme extends object>
   extends React.SFC<InnerProps & StyleProps & { theme?: Theme }>,
     ComponentSelector {
   /**
-   * @desc
-   * `Overwrapped` part checks the compatibility between `StyleProps` and `NewInnerProps`.
-   */
-  withComponent<NewInnerProps extends Overwrapped<StyleProps, NewInnerProps>>(
-    tag: React.ComponentType<NewInnerProps>
-  ): StyledComponent<NewInnerProps, StyleProps, Theme>
-
-  /**
-   * @desc this function overloading is unsafe
+   * @desc this method is type-unsafe
    */
-  withComponent<NewTag extends keyof JSX.IntrinsicElements>(
+  withComponent<NewTag extends keyof JSXInEl>(
     tag: NewTag
-  ): StyledComponent<JSX.IntrinsicElements[NewTag], StyleProps, Theme>
+  ): StyledComponent<JSXInEl[NewTag], StyleProps, Theme>
+  withComponent<Tag extends React.ComponentType<any>>(
+    tag: Tag
+  ): StyledComponent<PropsOf<Tag>, StyleProps, Theme>
 }
 
+type ReactClassPropKeys = keyof React.ClassAttributes<any>
 export interface CreateStyledComponentBase<
   InnerProps,
   ExtraProps,
   Theme extends object
 > {
   <
-    StyleProps extends Overwrapped<InnerProps, StyleProps> = InnerProps &
-      ExtraProps
+    StyleProps extends Omit<
+      Overwrapped<InnerProps, StyleProps>,
+      ReactClassPropKeys
+    > = Omit<InnerProps & ExtraProps, ReactClassPropKeys>
   >(
     ...styles: Array<Interpolation<WithTheme<StyleProps, Theme>>>
   ): StyledComponent<InnerProps, StyleProps, Theme>
   <
-    StyleProps extends Overwrapped<InnerProps, StyleProps> = InnerProps &
-      ExtraProps
+    StyleProps extends Omit<
+      Overwrapped<InnerProps, StyleProps>,
+      ReactClassPropKeys
+    > = Omit<InnerProps & ExtraProps, ReactClassPropKeys>
```

Comment:
@mitchellhamilton These part is about hiding `ref` and `key` from default `StyleProps`, so that
```jsx
const Comp = styled('div')({}).withComponent('label');
<Comp ref={(el: HTMLLabelElement) => {}} />
```
can make sense.
(Originally, `ref` of `Comp` was `((el: HTMLLabelElement) => void) & ((el: HTMLDivElement) => void)`.)